File: Trees/02_readTrees.cpp
#include <TFile.h>
#include <TTree.h>
#include <memory>
#include <vector>

using namespace std;

int main(int argc, char **argv) {
  if (argc == 1) {
    fprintf(stderr, "No file provided. Please run 01_saveTrees first\n");
    exit(1);
  }

  TFile *hfile = TFile::Open(argv[1], "READ");
  // or
  // TFile hfile(argv[1], "READ");

  // check that the file exist
  if (!hfile) {
    fprintf(stderr, "No file provided. Please run 01_saveTrees first\n");
    exit(1);
  }

  // Reading from the simple tree
  TTree *tree = nullptr;
  double px, py;
  int xx;

  hfile->GetObject("tree", tree);
  tree->SetBranchAddress("px", &px);
  tree->SetBranchAddress("py", &py);
  tree->SetBranchAddress("xx", &xx);
  size_t nEntries = (size_t)tree->GetEntries();

  for (size_t i = 0; i < nEntries; ++i) {
    tree->GetEntry(i);
    // Scientific code here ...
    // printf("%f, %f\n", px, py);
  }

  // Reading from the tree containing vectors
  TTree *vectree = nullptr;
  vector<double> *vec_px = nullptr;
  vector<double> *vec_py = nullptr;
  vector<int> *vec_xx = nullptr;
  hfile->GetObject("vectree", vectree);
  vectree->SetBranchAddress("px", &vec_px);
  vectree->SetBranchAddress("py", &vec_py);
  vectree->SetBranchAddress("xx", &vec_xx);
  nEntries = (size_t)vectree->GetEntries(); // just reusing the variable

  for (size_t i = 0; i < nEntries; ++i) {
    vectree->GetEntry(i);
    // More scientific code here ...
  }

  delete tree;
  delete vectree;
  delete hfile;
  return 0;
}

File: Trees/02_readTrees.py
import sys
from ROOT import TFile, TTree, std, AddressOf
import numpy as np
from array import array

def main(filename, which_tree, method="arr"):
    inputFile = TFile.Open(filename)

    if which_tree == "tree":
        tree = inputFile.Get("tree")
        nEntries = tree.GetEntries()
        if method == "arr":
            px = array('d', [0.])
            py = array('d', [0.])
            xx = array('i', [0])
            tree.SetBranchAddress("px", px)
            tree.SetBranchAddress("py", py)
            tree.SetBranchAddress("xx", xx)

            for i in range(0, nEntries):
                tree.GetEntry(i)
                # Scientific code here...
                print(px[0], py[0], xx[0])
        elif method == "numpy":
            px = np.array([0,], dtype=np.float)
            py = np.array([0.], dtype=np.float)
            xx = np.array([0], dtype=np.int)
            tree.SetBranchAddress("px", px)
            tree.SetBranchAddress("py", py)
            tree.SetBranchAddress("xx", xx)

            for i in range(0, nEntries):
                tree.GetEntry(i)
                # Scientific code here...
                print(px[0], py[0], xx[0])

    elif which_tree == "vectree":
        vectree = inputFile.Get("vectree")
        nEntries = vectree.GetEntries()

        if method == "arr":
            for _, event in enumerate(vectree):
                px = [a for a in event.px]
                py = [a for a in event.py]
                xx = [a for a in event.xx]
                # Scientific code here...


        elif method == "numpy":
            px = np.empty([nEntries, 10])
            py = np.empty([nEntries, 10])
            xx = np.empty([nEntries, 10])

            for i, event in enumerate(vectree):
                px[i] = [a for a in event.px]
                py[i] = [a for a in event.py]
                xx[i] = [a for a in event.xx]
                # Scientific code here...

        elif method == "std":
            px = std.vector("double")()
            py = std.vector("double")()
            xx = std.vector("int")()
            vectree.SetBranchAddress("px", px)
            vectree.SetBranchAddress("py", py)
            vectree.SetBranchAddress("xx", xx)

            for i in range(0, nEntries):
                vectree.GetEntry(i)
                # Scientific code here...
                # print(px, py, xx)

    inputFile.Close()

if __name__ == "__main__":
    main(sys.argv[1], sys.argv[2], sys.argv[3])
